src/Compilers/CSharp/Portable/FileBasedPrograms/VirtualProjectGenerator.cs

```diff
@@ -0,0 +1,452 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NET9_0_OR_GREATER
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.FileBasedPrograms;
+
+/// <summary>
+/// Can generate project file (XML) for a file-based program.
```

Comment:
> Would love an explanation (when this moves out of draft) of what a virtual project generator is. And what the ide would use it for.

There is some documentation here. It's for the 'dotnet run file' feature. It generates the behind-the-scenes virtual project. Currently this code lives in SDK. But IDE will need it for file-based program support as well.

Reply:
sorry, what i mean is "what is a virtual project" in the first place :)

Reply:
It's the project used to build file based programs. The project is like `.csproj` but only lives in memory. See https://github.com/dotnet/sdk/blob/main/documentation/general/dotnet-run-file.md for more details.

Reply:
FWIW, that link never mentions 'virtual project'.  It does mention 'implicit project' a bunch of times.

Reply:
Will fix, thanks